storage/src/adb/any/ordered/mod.rs

```diff
@@ -0,0 +1,843 @@
+
+        Ok(last_key)
+    }
+
+    /// Whether the span defined by `span_start` and `span_end` contains `key`.
+    pub fn span_contains(span_start: &O::Key, span_end: &O::Key, key: &O::Key) -> bool {
+        if span_start >= span_end {
+            // cyclic span case
+            if key >= span_start || key < span_end {
+                return true;
+            }
+        } else {
+            // normal span case
+            if key >= span_start && key < span_end {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Find the span produced by the provided `iter` that contains `key`, if any.
+    async fn find_span(
+        &self,
+        iter: impl Iterator<Item = &Location>,
+        key: &O::Key,
+    ) -> Result<Option<LocatedKey<O::Key, O::Value>>, Error> {
+        for &loc in iter {
+            // Iterate over conflicts in the snapshot entry to find the span.
+            let data = Self::get_update_op(&self.log, loc).await?;
+            if Self::span_contains(&data.key, &data.next_key, key) {
+                return Ok(Some((loc, data)));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Get the operation that defines the span whose range contains `key`, or None if the DB is
+    /// empty.
+    pub async fn get_span(
+        &self,
+        key: &O::Key,
+    ) -> Result<Option<LocatedKey<O::Key, O::Value>>, Error> {
+        if self.is_empty() {
+            return Ok(None);
+        }
+
+        // If the translated key is in the snapshot, get a cursor to look for the key.
+        let iter = self.snapshot.get(key);
+        let span = self.find_span(iter, key).await?;
+        if let Some(span) = span {
+            return Ok(Some(span));
+        }
+
+        let iter = self.snapshot.prev_translated_key(key);
+        let span = self.find_span(iter, key).await?;
+        if let Some(span) = span {
+            return Ok(Some(span));
+        }
+
+        // If we get here, then `key` must precede the first key in the snapshot, in which case we
+        // have to cycle around to the very last key.
+        let iter = self.snapshot.last_translated_key();
+        let span = self
+            .find_span(iter, key)
+            .await?
+            .expect("a span that includes any given key should always exist if db is non-empty");
+
+        Ok(Some(span))
+    }
+
+    /// Get the (value, next-key) pair of `key` in the db, or None if it has no value.
+    pub async fn get_all(&self, key: &O::Key) -> Result<Option<(O::Value, O::Key)>, Error> {
+        let Some(op) = self.get_key_op_loc(key).await?.map(|(op, _)| op) else {
+            return Ok(None);
+        };
+
+        let data = op
+            .into_key_data()
+            .expect("update operation must have key data");
+
+        Ok(Some((data.value, data.next_key)))
+    }
+
+    /// Whether the snapshot currently has no active keys.
+    pub fn is_empty(&self) -> bool {
+        self.active_keys == 0
+    }
```

Comment:
This is implemented both on the`IndexedLog` type itself and as a part of the `AnyDb` trait impl on `IndexedLog`; deduplicate?